gateway/src/dstack/gateway/core/store.py

```
@@ -0,0 +1,316 @@
+                stack.push_async_callback(supress_exc_async(run_async, ssh_tunnel.stop))
+
+                # Add to nginx
+                await self.nginx.add_upstream(
+                    service.domain, f"unix:{ssh_tunnel.sock_path}", replica.id
+                )
+                stack.push_async_callback(
+                    supress_exc_async(self.nginx.remove_upstream, service.domain, replica.id)
+                )
+
+                # All fine, remove rollbacks
+                stack.pop_all()
+
+            replica.ssh_tunnel = ssh_tunnel
+            service.replicas.append(replica)
+
+        logger.info(
+            "%s: replica %s for service %s (%s) is registered now",
+            project,
+            replica.id,
+            service_id,
+            service.domain,
+        )
+
+    async def unregister_replica(self, project: str, service_id: str, replica_id: str):
+        async with self._lock:
+            if service_id not in self.projects[project]:
+                raise GatewayError(
+                    f"Service ID {service_id!r} is not registered in project {project!r}"
+                )
+            service = self.services[service_id]
+
+            for replica in service.replicas:
+                if replica.id == replica_id:
+                    break
+            else:
+                raise GatewayError(
+                    f"Replica ID {replica_id!r} is not registered in service {service_id!r}"
+                )
+
+            logger.debug(
+                "%s: unregistering replica %s for service %s (%s)",
+                project,
+                replica_id,
+                service_id,
+                service.domain,
+            )
+
+            await asyncio.gather(
+                # Terminate SSH tunnel
+                run_async(replica.ssh_tunnel.stop),
+                # Remove from nginx
+                self.nginx.remove_upstream(service.domain, replica.id),
+                return_exceptions=True,
+            )
```

Same – ignored exceptions